### PR TITLE
Fix: Deploy to testnet script

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Script to deploy to testnets wasn't working `./deploy.sh`.

The problem was that dfx didn't know that the wasm was gzipped.

Solution: rename the wasm in dfx.json

More about the discussion: https://dfinity.slack.com/archives/CGA566TPV/p1686829378681289

# Changes
* [Commit 1](https://github.com/dfinity/nns-dapp/pull/2705/commits/88360aadbb3a180088b639db97b3aa24067bffe4):
  * Rename nns-dapp wasm in `dfx.json`.
  * Add new file extension to git ignore.
* [Commit2](https://github.com/dfinity/nns-dapp/pull/2705/commits/a270f06f5bec6aa049efdfa8462b45a1ac3251eb):  Make build-rs.sh emit .wasm.gz instead of .wasm (tested manually, it works)
* [Commit 3](https://github.com/dfinity/nns-dapp/pull/2705/commits/f10ef1d4fe80e182a5774b209f9552fe4561dd68): Update the dockerfile so that it too emits .wasm.gz (tested manually, it works)
* [Commit 4](https://github.com/dfinity/nns-dapp/pull/2705/commits/5ec7bf79c3b6949e98d1094b4c453de346f5b690): Bulk rename .wasm to .wasm.gz, but done manually checking every case as there are quite a few exceptions. E.g. external .wasm files are not renamed and jq paths ending in .wasm must not be renamed. (tested by CI; this verifies that users of the above changes within this codebase have been updated)
* [Commit 5](https://github.com/dfinity/nns-dapp/pull/2705/commits/199081aeee03d6ef6060a302254e5df13432a55e): For backwards compatibility, include a copy of all .wasm.gz assets with just the .wasm suffix.
* Non-PR change:  For recent releases, I have copied every .wasm in the release to the same path with suffix .wasm.gz.  See the comment in the MR for details.  This is to ensure that tasks such as rollback/rollforward tests have access to .wasm.gz for existing releases.  This may have to be repeated to cover any more releases created between now and the PR being merged.

# Tests
* There should be no change in functionality; existing CI should suffice for canister functionality.
* Here is a test release, to verify that both .wasm.gz and .wasm are produced: https://github.com/dfinity/nns-dapp/releases/tag/untagged-29888d6b4b9ee251e9b5
* Here are the last releases by proposal, to verify that .wasm.gz is available to rollback tests:
  * https://github.com/dfinity/nns-dapp/releases/tag/proposal-123011-agg
  * https://github.com/dfinity/nns-dapp/releases/tag/proposal-123006
  * https://github.com/dfinity/nns-dapp/releases/tag/proposal-122799

### CODE REVIEW

**Comment by bitdivine** [comment]
In principle I am fine with this.  The new suffix _is_ more accurate. However there are details I would like to check.

- One thing: I have deployed gzipped wasms with dfx 0.14.1 so I am curious why it is breaking now, and under what context.
- Currently docker-build creates nns-dapp.wasm and sns_aggregator.wasm and these are then copied to the root of the project.  At minimum, the copy would have to copy to the new location.  If we rename consistently everywhere, that will be quite a bit bigger as there are a few things that would have to be updated.

**Comment by bitdivine** [comment]
More fun:
```
+ dfx canister install --upgrade-unchanged sns_aggregator --wasm sns_aggregator_prod.wasm.gz --mode upgrade
Upgrading code for canister sns_aggregator, with canister ID bkyz2-fmaaa-aaaaa-qaaaq-cai
Error: Failed during wasm installation call: The Replica returned an error: code 5, message: "Wasm module of canister bkyz2-fmaaa-aaaaa-qaaaq-cai is not valid: Failed to decode wasm module: unsupported canister module format"
```
I have confirmed that this pulls the correct URL ( https://github.com/dfinity/nns-dapp/releases/download/aggregator-prod/sns_aggregator.wasm.gz ) and that teh file is a gzipped wasm.

**Comment by bitdivine** [comment]
Check a release to verify that the .wasm and .wasm.gz files are identical:
* Visited the release page.  Downloaded the .wasm and .wasm.gz files manually.
* Checked hashes:
```
max@sinkpad:~/Downloads (18)$ rm nns-dapp
nns-dapp-85863d3511556ab047e6f4ef9347ebc49af833bb.wasm     nns-dapp-85863d3511556ab047e6f4ef9347ebc49af833bb.wasm.gz  nns-dapp for mainnet.zip                                   nns-dapp_local.wasm                                        nns-dapp_t2.wasm                                           nns-dapp.wasm                                              
max@sinkpad:~/Downloads (18)$ rm nns-dapp*
max@sinkpad:~/Downloads (35)$ 
max@sinkpad:~/Downloads (35)$ 
max@sinkpad:~/Downloads (35)$ 
max@sinkpad:~/Downloads (36)$ sha256sum nns-dapp.wasm*
fd36071bcea746b1221a1e4ad4177527ded88c926ad560bd745a18eec915fd79  nns-dapp.wasm
fd36071bcea746b1221a1e4ad4177527ded88c926ad560bd745a18eec915fd79  nns-dapp.wasm.gz
max@sinkpad:~/Downloads (55)$ 
max@sinkpad:~/Downloads (1:03)$ 
max@sinkpad:~/Downloads (1:03)$ 
max@sinkpad:~/Downloads (1:03)$ rm sns_aggregator*
rm: cannot remove 'sns_aggregator': Is a directory
max@sinkpad:~/Downloads (1:12)$ sha256sum sns_aggregator.wasm*
1d4a74c431adde2bbe5239341f9b8353753f4a7671466283d66f65912858e9bc  sns_aggregator.wasm
1d4a74c431adde2bbe5239341f9b8353753f4a7671466283d66f65912858e9bc  sns_aggregator.wasm.gz
max@sinkpad:~/Downloads (1:33)$ 
```

File: build.sh [hunk 1]
```diff
@@ -14,6 +14,10 @@
 #                                                           │ ic-cdk-optimizer
 #                                                           │
 #                                                      nns-dapp.wasm
+#                                                           ▲
+#                                                           │ gzip
```
Comment (lmuntaner) [inline build.sh:18]: Thanks!

File: scripts/update-ii [hunk 2]
```diff
@@ -22,7 +22,7 @@ test -z "${unexpected_chars:-}" || {
 }
 
 : "Defining build script..."
-II_BUILD="$(printf 'curl --fail -sSL "%s" -o internet_identity_dev.wasm' "$II_URL")"
+II_BUILD="$(printf 'curl --fail -sSL "%s" -o internet_identity_dev.wasm.gz' "$II_URL")"
```
Comment (lmuntaner) [inline scripts/update-ii:25]: Did also II change to `wasm.gz`??

**Review by lmuntaner — COMMENTED** [review]
Thanks! I can't approve the PR because I started it, but LGTM.

I just added one question about II moving also to `.gz`

**Review by bitdivine — APPROVED** [review]
Note: I have made changes but these have been reviewed by @lmuntaner .

File: scripts/update-ii (hunk 2 above)
Comment (bitdivine) [reply scripts/update-ii:25]: Yes, see: https://github.com/dfinity/internet-identity/releases/tag/release-2023-06-16

This does mean that there is a backwards/forwards compatibility problem but we don't generally change II  version a lot so I don't expect this to be much of an issue.

Comment (bitdivine) [reply scripts/update-ii:25]: Closing the comment so that I can merge but happy to discuss further if wanted.

**Pull request merged**